### PR TITLE
Add an option to skip heal info check during brickEvict

### PR DESCRIPTION
**Pull request by iamniting**
### What does this PR achieve? Why do we need it?

This PR allows us to skip gluster volume heal info check while brickEvict or deviceRemove.

### CODE REVIEW

**Comment by phlogistonjohn** [comment]
I realize this is still WIP but the ci is failing for a legit reason, the api types has been modified but the uses in the functional test code are not updated to match.

**Comment by iamniting** [comment]
> I realize this is still WIP but the ci is failing for a legit reason: api types have been modified but the uses in the functional test code are not updated to match.

I have fixed all of them and CI is passed.

**Comment by iamniting** [comment]
retest this please

File: tests/functional/TestErrorHandling/tests/device_remove_test.go [hunk 1]
```diff
@@ -71,6 +72,127 @@ func TestDeviceRemoveOperation(t *testing.T) {
...
+		})
+		tests.Assert(t, err == nil, "expected err == nil, got:", err)
+
+		err = heketi.DeviceState(deviceToRemove, &api.StateRequest{
+			State:     api.EntryStateFailed,
+			HealCheck: api.HealCheckEnable,
+		})
+		tests.Assert(t,
+			strings.Contains(err.Error(), "i like to show fake error"),
+			"heketi was expected to ran gluster volume heal info but it did not ran it.")
```
Comment (phlogistonjohn) [inline tests/functional/TestErrorHandling/tests/device_remove_test.go:109]: "expected error triggered by 'gluster volume heal info' command, but no matching error was found"

File: tests/functional/TestErrorHandling/tests/device_remove_test.go [hunk 2]
```diff
@@ -71,6 +72,127 @@ func TestDeviceRemoveOperation(t *testing.T) {
...
+			State:     api.EntryStateOffline,
+			HealCheck: api.HealCheckEnable,
+		})
+		tests.Assert(t, err == nil, "expected err == nil, got:", err)
+
+		err = heketi.DeviceState(deviceToRemove, &api.StateRequest{
+			State:     api.EntryStateFailed,
+			HealCheck: api.HealCheckDisable,
+		})
+		tests.Assert(t, err == nil, "expected to not ran gluster volume info")
```
Comment (phlogistonjohn) [inline tests/functional/TestErrorHandling/tests/device_remove_test.go:151]: "expected no error from gluster commands, heal check disabled"?

**Comment by phlogistonjohn** [comment]
retest this please

**Comment by phlogistonjohn** [comment]
Go vet error:
INFO: changing to /var/tmp/gtg-wrapper/src/github.com/heketi/heketi
# github.com/heketi/heketi/apps/glusterfs
vet: apps/glusterfs/volume_entry_test.go:3309:4: cannot use api.EntryStateOffline (constant "offline" of type api.EntryState) as api.StateRequest value in argument to node.SetState

**Comment by phlogistonjohn** [comment]
Go test fails to compile. Possibly for simlar reasons as the vet error above.
Because travis ci is not functioning correctly, we need to run these tests by hand. Please run the "unit test suite" locally before pushing.

Eventually we'll either figure out what is wrong with travis ci or switch CIs. In the meantime, we need to do a little extra work ourselves. Sorry for the inconvenience.

File: apps/glusterfs/operations_device.go [hunk 3]
```diff
@@ -528,9 +528,13 @@ func (beo *BrickEvictOperation) execGetReplacmentInfo(
...
-	err = old.volume.canReplaceBrickInBrickSet(beo.db, executor, node, bs, index)
-	if err != nil {
-		return err
+	if beo.healCheck != api.HealCheckDisable {
+		err = old.volume.canReplaceBrickInBrickSet(beo.db, executor, node, bs, index)
+		if err != nil {
+			return err
+		}
+	} else {
+		logger.Info("skip heal info check for volume %+v", old.volume)
```
[the remark this answers is not in the document]
Comment (iamniting) [reply apps/glusterfs/operations_device.go]: done

File: client/cli/go/cmds/device.go [hunk 4]
```diff
@@ -43,6 +43,8 @@ func init() {
 		"Id of the node which has this device")
 	deviceAddCommand.Flags().Bool("destroy-existing-data", false,
 		"[DANGEROUS] Destroy any existing data on the device.")
+	deviceRemoveCommand.Flags().Bool("skip-heal", false,
```
[the remark this answers is not in the document]
Comment (iamniting) [reply client/cli/go/cmds/device.go]: done

File: client/cli/go/cmds/device.go [hunk 5]
```diff
@@ -163,6 +165,20 @@ var deviceRemoveCommand = &cobra.Command{
...
 
+		skipHeal, err := cmd.Flags().GetBool("skip-heal")
+		if err != nil {
+			return err
+		}
+
+		if skipHeal {
+			var option string
+			fmt.Println("Skip heal is dangerous and can lead to data loss of volume. Enter 'YES' to continue")
+			fmt.Scanln(&option)
```
[the remark this answers is not in the document]
Comment (iamniting) [reply client/cli/go/cmds/device.go]: done

File: client/cli/go/cmds/brick.go [hunk 6]
```diff
@@ -12,12 +12,15 @@ package cmds
...
 	"fmt"
 
+	"github.com/heketi/heketi/pkg/glusterfs/api"
 	"github.com/spf13/cobra"
 )
 
 func init() {
 	RootCmd.AddCommand(brickCommand)
 	brickCommand.AddCommand(brickEvictCommand)
+	brickEvictCommand.Flags().Bool("skip-heal", false,
```
[the remark this answers is not in the document]
Comment (iamniting) [reply client/cli/go/cmds/brick.go]: done

File: executors/cmdexec/volume.go [hunk 7]
```diff
@@ -425,12 +425,14 @@ func (s *CmdExecutor) HealInfo(host string, volume string) (*executors.HealInfo,
 	results, err := s.RemoteExecutor.ExecCommands(host, command,
 		s.GlusterCliExecTimeout())
 	if err := rex.AnyError(results, err); err != nil {
-		return nil, fmt.Errorf("Unable to get heal info of volume : %v", volume)
+		return nil, fmt.Errorf(
+			"Unable to get heal info of volume : %v with error: %v", volume, err)
```
[the remark this answers is not in the document]
Comment (iamniting) [reply executors/cmdexec/volume.go]: done

File: executors/cmdexec/volume.go [hunk 8]
```diff
@@ -425,12 +425,14 @@ func (s *CmdExecutor) HealInfo(host string, volume string) (*executors.HealInfo,
...
 	}
 	var healInfo CliOutput
 	err = xml.Unmarshal([]byte(results[0].Output), &healInfo)
 	if err != nil {
-		return nil, fmt.Errorf("Unable to determine heal info of volume : %v", volume)
+		return nil, fmt.Errorf(
+			"Unable to determine heal info of volume : %v with error: %v", volume, err)
```
[the remark this answers is not in the document]
Comment (iamniting) [reply executors/cmdexec/volume.go]: done

File: tests/functional/TestErrorHandling/tests/device_remove_test.go (hunk 1 above)
Comment (iamniting) [reply tests/functional/TestErrorHandling/tests/device_remove_test.go]: done

File: tests/functional/TestErrorHandling/tests/device_remove_test.go (hunk 2 above)
Comment (iamniting) [reply tests/functional/TestErrorHandling/tests/device_remove_test.go]: done

**Comment by iamniting** [comment]
> Go vet error:
> 
> ```
> vet: apps/glusterfs/volume_entry_test.go:3309:4: cannot use api.EntryStateOffline (constant "offline" of type api.EntryState) as api.StateRequest value in argument to node.SetState
> ```

I have tested an fixed all of them

**Comment by iamniting** [comment]
retest this please

**Comment by iamniting** [comment]
Thanks, @phlogistonjohn and @raghavendra-talur for reviews.

**Pull request merged**